compiler_opt/rl/corpus.py

```diff
@@ -32,11 +34,128 @@ class ModuleSpec:
   """Dataclass describing an input module and its compilation command options.
   """
   name: str
-  exec_cmd: Tuple[str, ...] = ()
-  size: int = 0
+  exec_cmd: Tuple[str, ...]
+  size: int
 
 
-class SamplerBucketRoundRobin:
+def _localize_cmdline(
+    orig_options: Tuple[str, ...],
+    context: Any = None,
+    additional_flags: Tuple[str, ...] = (),
+    delete_flags: Tuple[str, ...] = (),
+    replace_flags: Optional[Dict[str, str]] = None) -> List[str]:
+  option_iterator = iter(orig_options)
+  matched_replace_flags = set()
+  replace_flags = replace_flags if replace_flags is not None else {}
+
+  option = next(option_iterator, None)
+  cmdline = []
+  while option is not None:
+    if any(option.startswith(flag) for flag in delete_flags):
+      if '=' not in option:
+        next(option_iterator, None)
+    else:
+      matching_replace = [
+          flag for flag in replace_flags if option.startswith(flag)
+      ]
+      if not matching_replace:
+        cmdline.append(option)
+      else:
+        assert len(matching_replace) == 1
+        flag = matching_replace[0]
+        if flag in matched_replace_flags:
+          raise ValueError(f'{flag} was matched twice')
+        matched_replace_flags.add(flag)
+
+        if '=' not in option:
+          next(option_iterator, None)
+          cmdline.extend([option, replace_flags[flag].format(context=context)])
+        else:
+          cmdline.append(flag + '=' +
+                         replace_flags[flag].format(context=context))
+
+    option = next(option_iterator, None)
+  if len(matched_replace_flags) != len(replace_flags):
+    raise ValueError('flags that were expected to be replaced were not found')
+  cmdline.extend([f.format(context=context) for f in additional_flags])
+  return cmdline
+
+
+@dataclass(frozen=True)
+class LoadedModuleSpec:
+  """Encapsulates the loaded data of a module and the rules to persist it.
+
+  A LoadedModuleSpec can be passed to a remote location. There, given a local
+  directory, to_module_spec can be called, resulting in the data being saved
+  under that directory, the final compiler command line fully computed, and a
+  ready-to-use ModuleSpec returned.
+  """
+  name: str
+  module: bytes
+  thinlto_index: Optional[bytes] = None
+  orig_options: Tuple[str, ...] = ()
+  additional_flags: Tuple[str, ...] = ()
+  delete_flags: Tuple[str, ...] = ()
+  replace_flags: Optional[Dict[str, str]] = None
+
+  def _create_files_and_get_context(self, local_dir: str):
```

Comment:
It's not handling race conditions, intentionally. The caller is expected to create separate `local_dir`s (which it does). An alternative could be to let this function do that, but then it'd have to let the caller handle cleaning up (deleting the tempdir). In the end, the lifetime management of the directory is so much easier if left to the caller to fully handle in the first place.

Adding a comment though about the expectation from the caller. Adding it to `to_module_spec` because that's the public one.